In [115]:
import pandas as pd
import numpy as np

Reading CSV file with the help of Pandas library:

In [116]:
df = pd.read_csv("Bengaluru_House_Data.csv")

In [117]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [118]:
df.describe()

bath       balcony         price
count  13247.000000  12711.000000  13320.000000
mean       2.692610      1.584376    112.565627
std        1.341458      0.817263    148.971674
min        1.000000      0.000000      8.000000
25%        2.000000      1.000000     50.000000
50%        2.000000      2.000000     72.000000
75%        3.000000      2.000000    120.000000
max       40.000000      3.000000   3600.000000

In [119]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [120]:
df.shape

(13320, 9)

Now we check if there is some null value present or not by using isna() function:

In [121]:
df.isna().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

now we are applying value count on each column:

In [122]:
for column in df.columns:
    print(df[column].value_counts())
    print("*"*20)

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64
********************
Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
15-Aug               1
17-Jan               1
16-Nov               1
16-Jan               1
14-Jul               1
Name: availability, Length: 81, dtype: int64
********************
Whitefield                        540
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Abshot Layout                       1
Name: location, Length: 1305, dtype: int64
********************
2 BHK         5199
3 BHK        

so after seeing the data we will drop area_type,availabilty,society and balcony columns:

In [123]:
df.drop(columns=['area_type','availability','balcony','society'],inplace=True)

In [124]:
df.describe()

bath         price
count  13247.000000  13320.000000
mean       2.692610    112.565627
std        1.341458    148.971674
min        1.000000      8.000000
25%        2.000000     50.000000
50%        2.000000     72.000000
75%        3.000000    120.000000
max       40.000000   3600.000000

In [125]:
df.isna().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

**now we will fill and handel the Missing values:**


*starting with location*

In [126]:
df['location'].value_counts()

Whitefield                        540
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Abshot Layout                       1
Name: location, Length: 1305, dtype: int64

we can see here tha *White field* and *Sarjapur road* has the maximum number of counts:



as there is only one missing value we can fill it by any of the 2 max value because it wont cause many problems

in this case we will fill the null/missing value with *Sarjapur Road*.

In [127]:
df['location'] = df['location'].fillna('Sarjapur Road')

*Now lets handel missing values of size:*

In [128]:
df['size'].value_counts()

2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 BHK           2
11 Bedroom       2
10 BHK           2
14 BHK           1
13 BHK           1
12 Bedroom       1
27 BHK           1
43 Bedroom       1
16 BHK           1
19 BHK           1
18 Bedroom       1
Name: size, dtype: int64

As 2BHK has maximum value we will fill null/missing values with it

In [129]:
df['size'] = df['size'].fillna("2BHK")

*And lastly we will handle missing values of bath*

As the bath column is a numeric column we will fill the missing values with its median

In [130]:
df['bath'] = df['bath'].fillna(df['bath'].median())

In [131]:
df.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

so now we can see we have no null values hence we have sucesfully handeled missing values

In [132]:
df['size'].value_counts()

2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
2BHK            16
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
10 BHK           2
11 BHK           2
11 Bedroom       2
19 BHK           1
16 BHK           1
43 Bedroom       1
14 BHK           1
27 BHK           1
12 Bedroom       1
13 BHK           1
18 Bedroom       1
Name: size, dtype: int64

So here there is a problem, there are 2 ways to describe size of house i.e : BKH and bedroom

In [133]:
df['BHK'] = df['size'].str.split().str.get(0)

In [134]:
df['BHK'] = df['BHK'].replace('2BHK',2)

In [135]:
df['BHK'].value_counts()

2     5528
3     4857
4     1417
1      656
5      356
6      221
7      100
8       89
9       54
2       16
10      14
11       4
27       1
19       1
16       1
43       1
14       1
12       1
13       1
18       1
Name: BHK, dtype: int64

In [136]:
df['BHK'] = df['BHK'].astype(int)

In [137]:
df[df.BHK>20]

location        size total_sqft  bath  price  BHK
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

now in total_sqft there is the problem of range:


In [138]:
df['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

so to solv this problem we have to remove this range by taking mean of smallest and largest number

In [139]:
def convert_range(x):
    temp = x.split('-')
    if len(temp) == 2:
        return(float(temp[0])+float(temp[1])/2)
    try:
        return float(x)
    except :
        return None

In [140]:
df['total_sqft'] = df['total_sqft'].apply(convert_range)

In [141]:
df.head()

location       size  total_sqft  bath   price  BHK
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

now we will make a new column price_per_sqft which will help to remove outliers

In [142]:
df['price_per_sqft'] = df['price']*100000 / df['total_sqft']

In [143]:
df['price_per_sqft']

0         3699.810606
1         4615.384615
2         4305.555556
3         6245.890861
4         4250.000000
             ...     
13315     6689.834926
13316    11111.111111
13317     5258.545136
13318    10407.336319
13319     3090.909091
Name: price_per_sqft, Length: 13320, dtype: float64

In [144]:
df.describe()

total_sqft          bath         price           BHK  price_per_sqft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1572.225941      2.688814    112.565627      2.802778    7.883254e+03
std     1254.823072      1.338754    148.971674      1.294496    1.064303e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.227642e+03
50%     1280.000000      2.000000     72.000000      3.000000    5.416667e+03
75%     1690.000000      3.000000    120.000000      3.000000    7.293505e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

In [145]:
df['location'].value_counts()

Whitefield                        540
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Uvce Layout                         1
Abshot Layout                       1
Name: location, Length: 1306, dtype: int64

In [146]:
df['location'] = df['location'].apply(lambda x:x.strip())
location_count = df['location'].value_counts()

In [147]:
location_count

Whitefield                            541
Sarjapur  Road                        399
Electronic City                       304
Kanakpura Road                        273
Thanisandra                           237
                                     ... 
1Channasandra                           1
Hosahalli                               1
Vijayabank bank layout                  1
near Ramanashree California resort      1
Abshot Layout                           1
Name: location, Length: 1295, dtype: int64

As there are around 1295 types of location which is very large which we can not pass in the model hence we have to reduce it.


to reduce it we will take all the location which has value 10 or less than 10 as 'other'

In [148]:
location_count_less_10 = location_count[location_count<=10]
location_count_less_10

BTM 1st Stage                         10
Nagadevanahalli                       10
Basapura                              10
Sector 1 HSR Layout                   10
Dairy Circle                          10
                                      ..
1Channasandra                          1
Hosahalli                              1
Vijayabank bank layout                 1
near Ramanashree California resort     1
Abshot Layout                          1
Name: location, Length: 1054, dtype: int64

In [149]:
df['location'] = df['location'].apply(lambda x : 'other' if x in location_count_less_10 else x)

In [150]:
df['location'].value_counts()

other                 2886
Whitefield             541
Sarjapur  Road         399
Electronic City        304
Kanakpura Road         273
                      ... 
Nehru Nagar             11
Banjara Layout          11
LB Shastri Nagar        11
Pattandur Agrahara      11
Narayanapura            11
Name: location, Length: 242, dtype: int64

**Outlier detction and removal:**

In [151]:
df.describe()

total_sqft          bath         price           BHK  price_per_sqft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1572.225941      2.688814    112.565627      2.802778    7.883254e+03
std     1254.823072      1.338754    148.971674      1.294496    1.064303e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.227642e+03
50%     1280.000000      2.000000     72.000000      3.000000    5.416667e+03
75%     1690.000000      3.000000    120.000000      3.000000    7.293505e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

now lets find out what is total_sqft / BHK

In [152]:
(df['total_sqft']/df['BHK']).describe()

count    13274.000000
mean       579.732987
std        392.247238
min          0.250000
25%        475.000000
50%        554.000000
75%        627.500000
max      26136.000000
dtype: float64

so from here we can say that the houses which have total_sqft/BHK less than 300 is not fesible


hence we will remove that from the main datasest

In [153]:
df = df[((df['total_sqft']/df['BHK']))>=300]
df.describe()

total_sqft          bath         price           BHK  price_per_sqft
count  12530.000000  12530.000000  12530.000000  12530.000000    12530.000000
mean    1607.911902      2.559537    111.382401      2.650838     6278.292738
std     1277.977557      1.077938    152.077329      0.976678     4171.053803
min      300.000000      1.000000      8.440000      1.000000      267.829813
25%     1120.000000      2.000000     49.000000      2.000000     4177.777778
50%     1302.500000      2.000000     70.000000      3.000000     5273.281068
75%     1710.000000      3.000000    115.000000      3.000000     6896.551724
max    52272.000000     16.000000   3600.000000     16.000000   176470.588235

In [154]:
df.shape

(12530, 7)

In [155]:
def remove_outliers_sqft(dataframe):
    df_output = pd.DataFrame()
    for key,sub_df in dataframe.groupby('location'):
        m = np.mean(sub_df.price_per_sqft)

        st = np.std(sub_df.price_per_sqft)

        gen_df = sub_df[(sub_df.price_per_sqft > (m-st)) & (sub_df.price_per_sqft <=(m+st))]

        df_output = pd.concat([df_output,gen_df],ignore_index=True)
    return df_output

df = remove_outliers_sqft(df)
df.describe()

total_sqft         bath         price           BHK  price_per_sqft
count  10287.000000  10287.00000  10287.000000  10287.000000    10287.000000
mean    1517.904648      2.47176     91.305304      2.575872     5641.654913
std      903.645032      0.97517     86.528904      0.895390     2270.194350
min      300.000000      1.00000     10.000000      1.000000     1250.000000
25%     1110.000000      2.00000     49.000000      2.000000     4210.526316
50%     1290.000000      2.00000     67.000000      2.000000     5159.705160
75%     1650.000000      3.00000    100.000000      3.000000     6412.588374
max    30400.000000     16.00000   2200.000000     16.000000    24509.803922

here in this function :

it takes input as dataframe.
then we create an output dataframe
then we apply groupby with location, we know that when we apply groupby we will get a key and a sub dataframe from it.

so now in that sub dataframe we take the mean of price per sqt in variable m and std deviation of each location in variable st.

then we are filtering the data frame :

if the price pre sqft is between m-st and m+st then only we are keeping it in variable named gen_df

then we concat the gen_df and df_output in df_output ingnoring index.

and return df_output

then we apply this function on the main dataframe



In [156]:
def bhk_outlier_remover(dataframe):
    exclude_indices = np.array([])
    for location,location_df in dataframe.groupby('location'):
        bhk_stats = {}
        for bhk,bhk_df in location_df.groupby('BHK'):
            bhk_stats[bhk] = {
                'mean' : np.mean(bhk_df.price_per_sqft),
                'stddev' : np.std(bhk_df.price_per_sqft),
                'count' : bhk_df.shape[0]
            }
        for bhk,bhk_df in location_df.groupby('BHK'):
            stats = bhk_stats.get(bhk-1)
            if stats and stats['count']>5:
                exclude_indices = np.append(exclude_indices, bhk_df[bhk_df.price_per_sqft<(stats['mean'])].index.values)
    return dataframe.drop(exclude_indices,axis = 'index')



here in this function :

-->Again it takes dataframe as input

1) we are making an empty array named as exclude_indices it will store the indices which is not requried.
2) then we will apply groupby with location on dataframe from which we will get location and location_df.
3) then we made dictionary named bhk_stats which stores the stats of bhk in it.
4) then we will apply groupby with bhk on location_df from which we will get bhk and bhk_df.
5) now inside bhk_stats we will store 3 things 1) mean np.mean(bhk_df.price_per_sqft) 2) std np.std(bhk_df.price_per_sqft) 3) count bhk_df.shape[0].
6) after that process we will get information of each type of bhk at every location.
7) Again we will apply  groupby  with bhk on location_df.
8) here firstly we will check the values of bhk-1 and store it in variable named stats. if stats and stats['count'] > 5 then we will append in exclude_indices where bhk_df[bhk_df.price_per_sqft<(stats['mean'])].index.values.
9) we will drop exclude_indices from the dataframe



In [157]:
df = bhk_outlier_remover(df)

In [158]:
df.shape

(7358, 7)

In [159]:
df

location       size  total_sqft  bath  price  BHK  \
0      1st Block Jayanagar      4 BHK      2850.0   4.0  428.0    4   
1      1st Block Jayanagar      3 BHK      1630.0   3.0  194.0    3   
2      1st Block Jayanagar      3 BHK      1875.0   2.0  235.0    3   
3      1st Block Jayanagar      3 BHK      1200.0   2.0  130.0    3   
4      1st Block Jayanagar      2 BHK      1235.0   2.0  148.0    2   
...                    ...        ...         ...   ...    ...  ...   
10278                other      2 BHK      1200.0   2.0   70.0    2   
10279                other      1 BHK      1800.0   1.0  200.0    1   
10282                other      2 BHK      1353.0   2.0  110.0    2   
10283                other  1 Bedroom       812.0   1.0   26.0    1   
10286                other      4 BHK      3600.0   5.0  400.0    4   

       price_per_sqft  
0        15017.543860  
1        11901.840491  
2        12533.333333  
3        10833.333333  
4        11983.805668  
...               ...  
10278     5833.333333  
10279    11111.111111  
10282     8130.081301  
10283     3201.970443  
10286    11111.111111  

[7358 rows x 7 columns]

here 2 columns are not requried now : *Size and price_per_sqft* 

Therefore we will drop them.

In [160]:
df.drop(columns=['size','price_per_sqft'],inplace=True)

**Cleaned data**

In [161]:
df.head()

location  total_sqft  bath  price  BHK
0  1st Block Jayanagar      2850.0   4.0  428.0    4
1  1st Block Jayanagar      1630.0   3.0  194.0    3
2  1st Block Jayanagar      1875.0   2.0  235.0    3
3  1st Block Jayanagar      1200.0   2.0  130.0    3
4  1st Block Jayanagar      1235.0   2.0  148.0    2

now we will save this cleaned data in to another csv file named cleaned_data.csv

In [162]:
df.to_csv('cleaned_data.csv')

**Training the dataset :**

In [163]:
X = df.drop(columns = ['price'])
y = df['price']

importing all the requried modules for preprocessing:

In [164]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression,Lasso,Ridge
from sklearn.preprocessing import OneHotEncoder,StandardScaler
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.metrics import r2_score

In [165]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [166]:
print(X_train.shape)
print(X_test.shape)

(5886, 4)
(1472, 4)


Appling Linear Regration:


We will apply One hot Encoder on location because it is the only catagorical column in the data set

In [167]:
coloumn_trans = make_column_transformer((OneHotEncoder(sparse=False),['location']),remainder='passthrough')


In [168]:
scaler = StandardScaler()

In [173]:
lr  = LinearRegression()


In [174]:
pipe = make_pipeline(coloumn_trans,scaler,lr)

In [175]:
pipe.fit(X_train,y_train)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/preprocessing/_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  Fitted encoder.


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()),
                ('linearregression', LinearRegression())])

here whole data will firstly go to one hot encoder then it will be scaled and then the linear regression model will be applied

In [176]:
y_pred_lr = pipe.predict(X_test)

In [191]:
r2_score(y_test,y_pred_lr)*100

83.93664148380783

Now Applying Lasso:

In [178]:
lasso = Lasso()

In [185]:
pipe_1 = make_pipeline(coloumn_trans,scaler,lasso)

In [186]:
pipe_1.fit(X_train,y_train)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/preprocessing/_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  Fitted encoder.


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('lasso', Lasso())])

In [189]:
y_pred_lasso = pipe_1.predict(X_test)

In [192]:
r2_score(y_test,y_pred_lasso)*100

83.15444728453546

Now Applying Rigid:

In [193]:
rigid = Ridge()

In [194]:
pipe_2 = make_pipeline(coloumn_trans,scaler,rigid)

In [195]:
pipe_2.fit(X_train,y_train)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/preprocessing/_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  Fitted encoder.


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('ridge', Ridge())])

In [196]:
y_pred_rigid = pipe_2.predict(X_test)

In [197]:
r2_score(y_test,y_pred_rigid)*100

83.92909124513977

In [198]:
print("LinearRegression : ",r2_score(y_test,y_pred_lr)*100)
print("Lasso : ",r2_score(y_test,y_pred_lasso)*100)
print("Rigid : ",r2_score(y_test,y_pred_rigid)*100)

LinearRegression :  83.93664148380783
Lasso :  83.15444728453546
Rigid :  83.92909124513977


so from here we can see that Linear Regression is slightly better than rigid model hence we will choose Linear Regression model.

In [199]:
import pickle

In [200]:
pickle.dump(pipe,open('LRmodel.pkl','wb'))

Done with the model building here!!